File: src/com/anathema_roguelike/entities/characters/actions/TargetedAction.java
package com.anathema_roguelike.entities.characters.actions;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

import com.anathema_roguelike.entities.characters.Character;
import com.anathema_roguelike.entities.characters.actions.costs.ActionCosts;
import com.anathema_roguelike.entities.characters.actions.costs.EnergyCost;
import com.anathema_roguelike.entities.characters.perks.actions.ActionEffectModificationPerk;
import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.TargetEffect;
import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.Targetable;

public class TargetedAction<TargetType extends Targetable, OriginType extends Targetable> extends CharacterAction {
	
	private ArrayList<TargetEffect<? extends TargetType, ?>> targetEffects;
	
	private OriginType origin;
	private Collection<? extends TargetType> targets;
	
	@SafeVarargs
	public TargetedAction(Character character, OriginType origin, Collection<? extends TargetType> targets, EnergyCost energyCost, ActionCosts costs, TargetEffect<? extends TargetType, ?> ...targetEffects) {
		this(character, energyCost, costs, targetEffects);
		
		this.origin = origin;
		this.targets = targets;
	}
	
	@SafeVarargs
	public TargetedAction(Character character, EnergyCost energyCost, ActionCosts costs, TargetEffect<? extends TargetType, ?> ...targetEffects) {
		super(character, energyCost, costs);
		
		this.targetEffects = new ArrayList<>(Arrays.asList(targetEffects));
	}
	
	@SafeVarargs
	public TargetedAction(Character character, OriginType origin, Collection<? extends TargetType> targets, EnergyCost energyCost, TargetEffect<? extends TargetType, ?> ...targetEffects) {
		this(character, energyCost, targetEffects);
		
		this.origin = origin;
		this.targets = targets;
	}
	
	@SafeVarargs
	public TargetedAction(Character character, EnergyCost energyCost, TargetEffect<? extends TargetType, ?> ...targetEffects) {
		super(character, energyCost);
		
		this.targetEffects = new ArrayList<>(Arrays.asList(targetEffects));
	}
	
	public void setOrigin(OriginType origin) {
		this.origin = origin;
	}
	
	public void setTargets(Collection<? extends TargetType> targets) {
		this.targets = targets;
	}
	
	public OriginType getOrigin() {
		return origin;
	}
	
	public Collection<? extends TargetType> getTargets() {
		return targets;
	}
	
	@SuppressWarnings("unchecked")
	@Override
	protected void onTake() {
		if(origin == null || targets == null) {
			throw new RuntimeException("Invalid Targeted CharacterAction, Origin or Targets not set");
		}
		
		for(Targetable target : targets) {
			for(TargetEffect<? extends TargetType, ?> te : targetEffects) {
				if(te.getTargetType().isAssignableFrom(target.getClass())) {
					
					TargetEffect<? extends TargetType, ?> modified = te;
					for(ActionEffectModificationPerk<?, ?> aem : getActor().getPerks(ActionEffectModificationPerk.class)) {
						if(aem.getActionType().isAssignableFrom(getClass()) && te.getTargetType().isAssignableFrom(target.getClass())) {
							modified = aem.getClass().cast(aem).modify(this, modified);
						}
					}
					
					te.getClass().cast(te).applyTo(target);
				}
			}
		}
	}
	
	protected void addTargetEffect(TargetEffect<? extends TargetType, ?> te) {
		targetEffects.add(te);
	}
}

File: src/com/anathema_roguelike/entities/characters/actions/TargetedAction.scala
/*******************************************************************************
 * Copyright (c) 2019. Carl Minden
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 ******************************************************************************/

package com.anathema_roguelike
package entities.characters.actions


import com.anathema_roguelike.entities.characters.Character
import com.anathema_roguelike.entities.characters.actions.costs.{ActionCosts, EnergyCost}
import com.anathema_roguelike.entities.characters.perks.actions.targetingstrategies.{TargetEffect, Targetable}

class TargetedAction[TargetType <: Targetable](
    actor: Character,
    energyCost: EnergyCost,
    additionalCosts: ActionCosts,
    targetEffects: TargetEffect[TargetType, _]*
  ) extends CharacterAction(actor, energyCost, additionalCosts.getCosts.toSeq: _*) {

  var targets: Iterable[_ <: TargetType] = List()

  def getTargets: Iterable[_ <: TargetType] = targets
  def setTargets(targets: Iterable[_ <: TargetType]): Unit = this.targets = targets

  override protected def onTake(): Unit = {
    if (targets.isEmpty) throw new RuntimeException("Invalid Targeted CharacterAction, Targets not set")

    for (target <- targets) {
      for (te <- targetEffects) {
        if (te.getTargetType.isAssignableFrom(target.getClass)) {
          te.applyTo(target)
        }
      }
    }
  }
}
